File: 🟡 Medium/2938 - Separate Black and White Balls/_2938SeparateBlackandWhiteBalls.java
class Solution {
    public long minimumSteps(String s) {
        int zero = s.length() - 1;
        int one = s.length() - 1;

        StringBuilder sb = new StringBuilder(s);

        while (one >= 0 && sb.charAt(one) == '1') {
            one--;
            zero--;
        }

        long steps = 0;

        while (one >= 0) {
            while (one >= 0 && sb.charAt(one) != '1') {
                one--;
            }

            while (zero >= 0 && sb.charAt(zero) != '0') {
                zero--;
            }

            if (zero == -1 || one == -1) {
                break;
            }

            steps += zero - one;

            sb.setCharAt(one, '0');
            sb.setCharAt(zero, '1');
        }

        return steps;

    }
}

File: 🟡 Medium/2938 - Separate Black and White Balls/_2938SeparateBlackandWhiteBalls.js
/**
 * @param {string} s
 * @return {number}
 */
var minimumSteps = function (s) {
    let res = 0;
    let whitePos = 0;
    let zero = 0;

    while (zero < s.length) {
        if (s[zero] == '0') {
            res += (zero - whitePos);
            whitePos++;
        }
        zero++;
    }

    return res;
};